airbyte-integrations/connectors/destination-s3/src/main/java/io/airbyte/integrations/destination/s3/S3StorageOperations.java

```diff
@@ -59,7 +60,30 @@ public String getBucketObjectPath(final String namespace, final String streamNam
             .replaceAll(Pattern.quote("${HOUR}"), String.format("%02d", writeDatetime.hourOfDay().get()))
             .replaceAll(Pattern.quote("${MINUTE}"), String.format("%02d", writeDatetime.minuteOfHour().get()))
             .replaceAll(Pattern.quote("${SECOND}"), String.format("%02d", writeDatetime.secondOfMinute().get()))
+            .replaceAll(Pattern.quote("${MILLISECOND}"), String.format("%02d", writeDatetime.millisOfSecond().get()))
+            .replaceAll(Pattern.quote("${EPOCH}"), String.format("%d", writeDatetime.getMillis()))
+            .replaceAll(Pattern.quote("${UUID}"), String.format("%s", UUID.randomUUID()))
             .replaceAll("/+", "/"));
+    if (customPathFormat.contains("${PART_ID}")) {
+      return objectPath.replaceAll(Pattern.quote("${PART_ID}"), String.format("%d", countObjectsInBucket(objectPath)));
```

Comment:
Suggestion: track the part ID inside this class, instead of querying the bucket each time. That will make unit test slightly easier.

Reply:
But it's possible that files already exist in the bucket or some other process may write in parallel?
Thus tracking the number of files in the object may not be accurate enough?

Reply:
Got it.